🤖 Machine Learning &nbsp;·&nbsp; *Bài 17/23*

[« Machine Learning - Logistic Regression](016_ml_logistic_regression.ipynb) &nbsp;•&nbsp; [📚 Mục lục](../README.md) &nbsp;•&nbsp; [Machine Learning - Preprocessing - Categorical Data »](018_ml_categorical_data.ipynb)

# Machine Learning - Grid Search

> 📖 **Học song song trên W3Schools:** [python_ml_grid_search.asp](https://www.w3schools.com/python/python_ml_grid_search.asp)
>
> 🎯 **Trong bài này:** Grid Search · How does it work? · Using Default Parameters · Implementing Grid Search · Results Explained · Note on Best Practices · ➕ Mở rộng: GridSearchCV

## Grid Search

Phần lớn mô hình học máy có các **tham số** có thể điều chỉnh để thay đổi cách mô hình học. Ví dụ, mô hình hồi quy logistic trong `sklearn` có tham số **`C`** điều khiển mức **regularization** (chính quy hoá) — ảnh hưởng tới độ phức tạp của mô hình.

Làm sao chọn giá trị tốt nhất cho `C`? Giá trị tốt nhất phụ thuộc vào dữ liệu dùng để huấn luyện.

## How does it work? — Hoạt động thế nào?

Một cách là **thử nhiều giá trị khác nhau** rồi chọn giá trị cho điểm số tốt nhất. Kỹ thuật này gọi là **grid search**. Nếu cần chọn hai hay nhiều tham số, ta thử mọi tổ hợp — tạo thành một **lưới (grid)** giá trị.

Trước khi làm ví dụ, cần hiểu tham số ta đang thay đổi: giá trị `C` **cao** báo cho mô hình rằng dữ liệu huấn luyện **giống thực tế**, nên đặt trọng số lớn hơn vào dữ liệu huấn luyện. Giá trị `C` **thấp** làm ngược lại.

## Using Default Parameters — Dùng tham số mặc định

Xem kết quả khi **không** grid search, chỉ dùng tham số mặc định. Nạp tập dữ liệu **iris** có sẵn trong sklearn:

In [1]:
from sklearn import datasets

iris = datasets.load_iris()

X = iris['data']
y = iris['target']

Tạo mô hình với `max_iter` lớn để mô hình đủ vòng lặp tìm kết quả:

In [2]:
from sklearn.linear_model import LogisticRegression

logit = LogisticRegression(max_iter=10000)

print(logit.fit(X, y))
print(logit.score(X, y))

LogisticRegression(max_iter=10000)
0.9733333333333334


Với `C = 1` mặc định, điểm số là khoảng **0.973**.

## Implementing Grid Search — Thực hiện grid search

Tạo danh sách các giá trị `C` cần thử, rồi lặp qua từng giá trị:

In [3]:
C = [0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2]

scores = []

for choice in C:
    logit.set_params(C=choice)
    logit.fit(X, y)
    scores.append(logit.score(X, y))

print(scores)

[0.9666666666666667, 0.9666666666666667, 0.9733333333333334, 0.9733333333333334, 0.98, 0.98, 0.9866666666666667, 0.9866666666666667]


## Results Explained — Giải thích kết quả

Các giá trị `C` **thấp** cho kết quả kém hơn giá trị mặc định. Khi tăng `C` lên **1.75**, độ chính xác tăng lên.

Tăng `C` thêm nữa không giúp ích — giá trị này coi như đạt tối đa.

## Note on Best Practices — Lưu ý thực hành tốt

Ta đã chấm điểm mô hình **trên chính dữ liệu huấn luyện**. Nếu mô hình bám quá sát dữ liệu này, nó có thể **không tổng quát hoá** tốt với dữ liệu chưa thấy — gọi là **overfitting**.

Để tránh bị đánh lừa bởi điểm số trên tập huấn luyện, nên tách riêng một phần dữ liệu để **đánh giá** mô hình — xem bài [Train/Test](012_ml_train_test.ipynb) và [Cross Validation](021_ml_cross_validation.ipynb).

## ➕ Mở rộng: GridSearchCV

scikit-learn có sẵn `GridSearchCV` — kết hợp grid search với cross validation:

In [4]:
from sklearn.model_selection import GridSearchCV

search = GridSearchCV(LogisticRegression(max_iter=10000), {"C": C}, cv=5)
search.fit(X, y)
print(search.best_params_, round(search.best_score_, 3))

{'C': 1} 0.973


---

## 🧠 Ghi nhớ nhanh

> - Grid search = thử **mọi giá trị/tổ hợp** tham số, chọn cái cho điểm tốt nhất.
> - Tham số `C` của LogisticRegression điều khiển regularization.
> - Đánh giá trên dữ liệu **chưa thấy** để tránh overfitting → `GridSearchCV(..., cv=5)`.

---

## 📝 Exercise — Bài tập nhanh

**❓ Câu hỏi:** Grid search dùng để làm gì?

- **A.** Chia dữ liệu train/test
- **B.** Vẽ lưới trên biểu đồ
- **C.** Chuẩn hoá dữ liệu
- **D.** Tìm giá trị tham số tốt nhất cho mô hình

<details>
<summary>👉 Xem đáp án</summary>

✅ **Đáp án: D** — Tìm giá trị tham số tốt nhất cho mô hình

</details>

---

[« Machine Learning - Logistic Regression](016_ml_logistic_regression.ipynb) &nbsp;•&nbsp; [📚 Mục lục](../README.md) &nbsp;•&nbsp; [Machine Learning - Preprocessing - Categorical Data »](018_ml_categorical_data.ipynb)